orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Stage.java

```diff
@@ -384,81 +369,76 @@ public Task taskById(String taskId) {
 
-    if (!requisiteStageRefIds.isEmpty() && !directParentOnly) {
-      // Get stages this stage depends on via requisiteStageRefIds:
-      List<Stage> previousStages =
-          execution.getStages().stream()
-              .filter(it -> requisiteStageRefIds.contains(it.refId))
-              .filter(it -> !visited.contains(it.refId))
-              .collect(toList());
-      List<Stage> syntheticStages =
-          execution.getStages().stream()
-              .filter(
-                  s ->
-                      previousStages.stream()
-                          .map(Stage::getId)
-                          .anyMatch(id -> id.equals(s.parentStageId)))
-              .collect(toList());
-      return ImmutableList.<Stage>builder()
-          .addAll(previousStages)
-          .addAll(syntheticStages)
-          .addAll(
-              previousStages.stream()
-                  .flatMap(it -> it.getAncestorsImpl(visited, directParentOnly).stream())
-                  .collect(toList()))
-          .build();
-    } else if (parentStageId != null && !visited.contains(parentStageId)) {
-      // Get parent stages, but exclude already visited ones:
-
-      List<Stage> ancestors = new ArrayList<>();
-      if (getSyntheticStageOwner() == SyntheticStageOwner.STAGE_AFTER) {
-        ancestors.addAll(
+    toVisit.add(this);
+
+    while (!toVisit.isEmpty()) {
+      Stage curStage = toVisit.remove();
+
+      if (visited.contains(curStage.id)) {
+        continue;
+      }
+
+      visited.add(curStage.id);
+      allAncestorStages.add(curStage);
+
+      // 1. Find all stages we depend on through requisiteStageRefIds and push them on the queue
+      if (!curStage.requisiteStageRefIds.isEmpty() && !directParentOnly) {
+        List<Stage> refIdPriorStages =
             execution.getStages().stream()
-                .filter(
-                    it ->
-                        parentStageId.equals(it.parentStageId)
-                            && it.getSyntheticStageOwner() == SyntheticStageOwner.STAGE_BEFORE)
-                .collect(toList()));
+                .filter(it -> curStage.requisiteStageRefIds.contains(it.refId))
+                .collect(toList());
+
+        toVisit.addAll(refIdPriorStages);
       }
 
-      ancestors.addAll(
-          execution.getStages().stream()
-              .filter(it -> it.id.equals(parentStageId))
-              .findFirst()
-              .<List<Stage>>map(
-                  parent ->
-                      ImmutableList.<Stage>builder()
-                          .add(parent)
-                          .addAll(parent.getAncestorsImpl(visited, directParentOnly))
-                          .build())
-              .orElse(emptyList()));
-
-      return ancestors;
-    } else {
-      return emptyList();
+      // 2. If we have a parent stage
+      if (curStage.parentStageId != null && !visited.contains(curStage.parentStageId)) {
+        List<Stage> ancestors = new ArrayList<>();
```

Comment:
Any reason to have this temporary variable vs. just directly doing `toVisit.addAll()` and `toVisit.add()` below?  (I'm asking primarily for clarity/readability, though I suppose it would probably also be a bit faster to remove it as well.)

Reply:
i was trying to be clearer, but you're right that it would probably be faster and generate less GC cruft. let me try it with the immutablelist builder